### PR TITLE
Api 6218 charon prep

### PR DESCRIPTION
**Pull request by crolarlibertyva**
## Description of change
Adds prescreening of the token payload that occurs prior to the call to the Charon service.

## Original issue(s)
https://vajira.max.gov/browse/API-6218

## Things to know about this PR

- This update is added to V2 validation. This change does not include the call to the Charon service yet.
- Related to https://github.com/department-of-veterans-affairs/devops/pull/9074

### CODE REVIEW

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (beaugrantham-va) [inline modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:104]: We're told that _most_ users will only have a single vista_id, but it's possible for there to be more than one (as with the test user). I think we need to check them all.

The direction in the story (and other conversations) may have not implied this, and it was probably my bad.

> Ensure that `act.vista_id` contains an entry for `launch.sta3n`

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (beaugrantham-va) [inline modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:14]: Is this necessary? I don't recall seeing this pattern for other 4xx errors.

Comment (crolarlibertyva) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:14]: The endpoint returned 500 if I didn't add the line.

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (crolarlibertyva) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:104]: It was clear and forgot to finish a redo. My original implementation covered this but it was much looser check.

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (crolarlibertyva) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb]: I took out the rescue altogether to address this.

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (lukemarkwordtlibertyits) [inline modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:91]: This is just a preference so take it or leave it, but i think vista_id is short enough where the abbreviation to vid is not necessary.

**Review by lukemarkwordtlibertyits — CHANGES_REQUESTED** [review]

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (crolarlibertyva) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:93]: There has to be launch scope for those token issued for certain particular audiences. because the launch context is where the station id (used as a charon parameter) comes from.  Not all ssoi tokens will need additional validation. But any ssoi token with an audience flagged as one that should need additional charon validation is not authorized if it doesn't have a launch context.

Comment (lukemarkwordtlibertyits) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:93]: Ok so in that case the response will be as follows:

```json
{"errors":[{"title":"Not authorized","detail":"Invalid request","code":"401","status":"401"}]}
```

By saying "not all ssoi tokens will need additional validation" you are saying they will not hit the validation endpoint?

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (crolarlibertyva) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:91]: Will change to make more readable.

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (crolarlibertyva) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:93]: no. but rather not go through additional checks, using charon during validation.

**Comment by lukemarkwordtlibertyits** [comment]
Some regression Tests I have ran on this branch:

```sh
 ✓ Valid Token. Patient Scp. Valid ICN. Should return 200
 ✓ Valid Token. With Launch/Patient scp. Patient Scp. Valid ICN. Should return 200
 ✓ Valid Token. Patient Scp. Invalid ICN. Should return !200
 ✓ Valid Token. With Launch/Patient scp. Patient Scp. Invalid ICN. Should return !200
 ✓ Patient Token.  Valid. With Launch/Patient scp. Return 200 with appropriate fields
✓ SSOI Token. Valid. With Launch scp. Return 200 with appropriate fields
```

**Review by lukemarkwordtlibertyits — DISMISSED** [review]
Approved assuming the following Specs are desired:

1. SSOI Token with Launch Scope and Valid Charon Audience. 
	- Returns valid token response.
2. SSOI Token with Launch Scope and Invalid Charon Audience.
	- Returns valid token response.
3. SSOI Token without Launch Scope and valid Charon Audience.
	- returns 401
```json
{"errors":[{"title":"Not authorized","detail":"Invalid request","code":"401","status":"401"}]}
```

4. SSOI Token withouth Launch Scope and invalid Charon Audience.
	- Returns valid token response.

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (beaugrantham-va) [inline modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:99]: Let me reach out to the team that provided me this regex and ask about valid `sta3n` values.

They didn't provide regex groupings, but assuming that `|` and `^` are delimiters, I'm wondering if this regex should actually be `\d{3}[A-Z]*`.

I'll let you know.

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (beaugrantham-va) [inline modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:101]: Are these two lines (100, 101) driven by linting rules? The double inverted logic is quite hard to read. I had to break out `irb` to make sure I understood this.

`return true if sta3n.to_s.eql?(parsed_sta3n.to_s)` is a lot cleaner.

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (beaugrantham-va) [inline modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:90]: Can this be renamed to a boolean style method (or docs added, or both) to make it clear what to expect from this?

What does it mean for a value of `true` or `false` to be returned? I had to dig into the logic to understand it.

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (beaugrantham-va) [inline modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:14]: I'd like to get @dbrown-va to get some eyes on this. I'm not sure how this affects unhandled/runtime errors.

**Comment by beaugrantham-va** [comment]
Can you open and link a PR for the real config changes?

* Enabled in staging, prod
* Disabled dev, sandbox

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (crolarlibertyva) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:99]: I will change to make sure. It's safer to go with your recommendation for now and I'll adjust per feedback.

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (crolarlibertyva) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:14]: This part is changed in in the part2 changes.

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (crolarlibertyva) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb]: The change I made doesn't include the last "\|" since so the parsed section wouldn't include that delimiter.

File: modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb
Comment (dbrown-va) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:14]: I think this removal will allow for unexpected exceptions to be raised in a less controlled fashion.

Earlier in the PR thread I see there was the `e.detail.eql?('Invalid request')` logic inside the rescue block.

Checking out the branch now to take a look at the different scenarios that could occur.

Comment (crolarlibertyva) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:14]: That original one would need to check e.detail.is_nil? if it were still there.

Comment (dbrown-va) [reply modules/openid_auth/app/controllers/openid_auth/v2/validation_controller.rb:14]: I think checking for the `Common::Exceptions::TokenValidationError` error type which is set with the `error_klass` method would provide the most consistency.  Then ensure any other unexpected errors are returned at least as the InternalServerError rather than a stack trace.

**Review by dbrown-va — DISMISSED** [review]
I think there is still opportunity to clean up the re-raised exception to prevent stack traces returned in the response, but that was existing prior to this charon addition.

Should not merge until the config PR is approved and merged.

**Review by dbrown-va — APPROVED** [review]
Confirmed Devops PR for settings was merged.

**Pull request merged**
